src/callbacks/steady_state.jl

```diff
@@ -0,0 +1,163 @@
+    previous_ekin :: Vector{ELTYPE}
+    interval_size :: Int
+end
+
+function SteadyStateCallback(; interval::Integer=0, dt=0.0, interval_size::Integer=10,
+                             abstol=1.0e-8, reltol=1.0e-6)
+    abstol, reltol = promote(abstol, reltol)
+
+    if dt > 0 && interval > 0
+        throw(ArgumentError("setting both `interval` and `dt` is not supported"))
+    end
+
+    if dt > 0
+        interval = Float64(dt)
+    end
+
+    steady_state_callback = SteadyStateCallback(interval, abstol, reltol, [Inf64],
+                                                interval_size)
+
+    if dt > 0
+        return PeriodicCallback(steady_state_callback, dt, save_positions=(false, false),
+                                final_affect=true)
+    else
+        return DiscreteCallback(steady_state_callback, steady_state_callback,
+                                save_positions=(false, false))
+    end
+end
+
+function Base.show(io::IO, cb::DiscreteCallback{<:Any, <:SteadyStateCallback})
+    @nospecialize cb # reduce precompilation time
+
+    cb_ = cb.affect!
+
+    print(io, "SteadyStateCallback(abstol=", cb_.abstol, ", ", "reltol=", cb_.reltol, ")")
+end
+
+function Base.show(io::IO,
+                   cb::DiscreteCallback{<:Any,
+                                        <:PeriodicCallbackAffect{<:SteadyStateCallback}})
+    @nospecialize cb # reduce precompilation time
+
+    cb_ = cb.affect!.affect!
+
+    print(io, "SteadyStateCallback(abstol=", cb_.abstol, ", reltol=", cb_.reltol, ")")
+end
+
+function Base.show(io::IO, ::MIME"text/plain",
+                   cb::DiscreteCallback{<:Any, <:SteadyStateCallback})
+    @nospecialize cb # reduce precompilation time
+
+    if get(io, :compact, false)
+        show(io, cb)
+    else
+        cb_ = cb.affect!
+
+        setup = ["absolute tolerance" => cb_.abstol,
+            "relative tolerance" => cb_.reltol,
+            "interval" => cb_.interval,
+            "interval size" => cb_.interval_size]
+        summary_box(io, "SteadyStateCallback", setup)
+    end
+end
+
+function Base.show(io::IO, ::MIME"text/plain",
+                   cb::DiscreteCallback{<:Any,
+                                        <:PeriodicCallbackAffect{<:SteadyStateCallback}})
+    @nospecialize cb # reduce precompilation time
+
+    if get(io, :compact, false)
+        show(io, cb)
+    else
+        cb_ = cb.affect!.affect!
+
+        setup = ["absolute tolerance" => cb_.abstol,
+            "relative tolerance" => cb_.reltol,
+            "interval" => cb_.interval,
+            "interval_size" => cb_.interval_size]
+        summary_box(io, "SteadyStateCallback", setup)
+    end
+end
+
+# `affect!` (`PeriodicCallback`)
+function (cb::SteadyStateCallback)(integrator)
+    steady_state_condition!(cb, integrator) || return nothing
+
+    print_summary(integrator)
+
+    # `terminate!(integrator)` terminates the simulation immediately and might cause an error message.
+    integrator.opts.maxiters = integrator.iter
```

Comment:
This causes a warning message.
A more elegant solution would be to initially (in `initialize!`)  sort the `CallbackSet` in such a way that the `SteadyStateCallback` is last.  
The set is stored in `integrator.opts.callback`. However, I have no idea how to check if the actual callback is the last element or not.

Any idea @efaulhaber  and @svchb ?

Reply:
Another (simpler) solution:
`@suppress integrator.opts.maxiters = integrator.iter`

from https://github.com/JuliaIO/Suppressor.jl

Reply:
I guess the only way would be to implement a custom solve function which wraps the ODE solve?

Reply:
My prefered approach is something like:

```julia
function initialize_cb!(cb::DiscreteCallback{<:Any,
                                             <:PeriodicCallbackAffect{<:SteadyStateCallback}},
                        u, t, integrator)
    callback_set = integrator.opts.callback.discrete_callbacks

    last(callback_set) isa typeof(cb) && return nothing

    # Make sure that the last callback is a `SteadyStateCallback` to avoid an error message when terminating
    new_set = (filter(i -> !(i isa typeof(cb), callback_set))...,
               filter(i -> i isa typeof(cb), callback_set)...)

    integrator.opts.callback.discrete_callbacks = new_set

    return nothing
end
```

However, I think we can't just mutate the `CallbackSet`.

@efaulhaber what's your opinion?

Reply:
Otherwise a solution would be to adjust the CallBacks upstream?

Reply:
Why not use `terminate!`? This is the recommended option:
https://docs.sciml.ai/DiffEqDocs/stable/features/callback_functions/#Example-2:-Terminating-an-Integration
<img width="850" alt="grafik" src="https://github.com/user-attachments/assets/95b88aaf-abac-4e93-b2d8-de8949c83fc5">

Reply:
See comment in code above

Reply:
`terminate!` was fixed in https://github.com/SciML/DiffEqCallbacks.jl/pull/229.
This is the better solution, which does show the misleading warning.
However, before we can use the latest version of DiffEqCallbacks, we first have to wait for https://github.com/JuliaSIMD/StrideArrays.jl/issues/92.